snorkel/learning/pytorch/noise_aware_model.py

```diff
@@ -0,0 +1,248 @@
+        else:
+            raise Exception("[{0}] No model found at <{1}>".format(
+                self.name, model_name))
+
+    def save(self, model_name=None, save_dir='checkpoints', verbose=True,
+        global_step=0):
+        """Save current model."""
+        model_name = model_name or self.name
+
+        # Note: Model checkpoints need to be saved in separate directories!
+        model_dir = os.path.join(save_dir, model_name)
+        if not os.path.exists(model_dir):
+            os.makedirs(model_dir)
+
+        torch.save((self.cardinality, self.name, self.model_kwargs), '{}/model.kwargs'.format(model_dir))
+        torch.save(self.state_dict(), '{}/model.params'.format(model_dir))
+
+        if verbose:
+            print("[{0}] Model saved as <{1}>".format(self.name, model_name))
+        
+    def train(self, X_train, Y_train, n_epochs=25, lr=0.01, batch_size=64,
+        rebalance=False, X_dev=None, Y_dev=None, print_freq=1, dev_ckpt=True,
+        dev_ckpt_delay=0.75, save_dir='checkpoints', seed=123, **kwargs):
+        """
+        Generic training procedure for PyTorch model
+
+        :param X_train: The training Candidates. If self.representation is True, then
+            this is a list of Candidate objects; else is a csr_AnnotationMatrix
+            with rows corresponding to training candidates and columns 
+            corresponding to features.
+        :param Y_train: Array of marginal probabilities for each Candidate
+        :param n_epochs: Number of training epochs
+        :param lr: Learning rate
+        :param batch_size: Batch size for SGD
+        :param rebalance: Bool or fraction of positive examples for training
+                    - if True, defaults to standard 0.5 class balance
+                    - if False, no class balancing
+        :param X_dev: Candidates for evaluation, same format as X_train
+        :param Y_dev: Labels for evaluation, same format as Y_train
+        :param print_freq: number of epochs at which to print status, and if present,
+            evaluate the dev set (X_dev, Y_dev).
+        :param dev_ckpt: If True, save a checkpoint whenever highest score
+            on (X_dev, Y_dev) reached. Note: currently only evaluates at
+            every @print_freq epochs.
+        :param dev_ckpt_delay: Start dev checkpointing after this portion
+            of n_epochs.
+        :param save_dir: Save dir path for checkpointing.
+        :param kwargs: All hyperparameters that change how the network is built
+            must be passed through here to be saved and reloaded to save /
+            reload model. *NOTE: If a parameter needed to build the 
+            network and/or is needed at test time is not included here, the
+            model will not be able to be reloaded!*
+        """
+        # Sets random seeds
+        torch.manual_seed(seed)
+        torch.cuda.manual_seed(seed)
+        torch.cuda.manual_seed_all(seed)
```

Comment:
See comment above- might want to add in option to set CuDNN off in case that causes non-determinism?

Reply:
Yeah I agree, good point